File: java/src/main/java/com/aquidify/sdk/Aquidify.java
package com.aquidify.sdk;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.io.IOException;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;

/**
 * Aquidify API client.
 *
 * <pre>{@code
 * var aq = new Aquidify(System.getenv("AQUIDIFY_API_KEY"));
 * var r = aq.interpret("hiring.candidate", "Iščem delo v skladišču v Ljubljani, brez nočnih.", "sl-SI");
 * r.interpretation().at("/intents/0/roles/0/value").asText(); // "warehouse"
 *
 * // your own task, any industry: register it once, then interpret with its id
 * aq.putTask("support.ticket@1.0.0", Map.of(
 *     "instructions", "Classify customer support emails.",
 *     "schema", Map.of("type", "object", "properties", Map.of(
 *         "category", Map.of("enum", List.of("billing", "bug"))))));
 * aq.interpret("support.ticket@1.0.0", body, "en");
 * }</pre>
 */
public final class Aquidify {
    public static final String VERSION = "0.2.0";

    private static final ObjectMapper JSON = new ObjectMapper();
    private final String apiKey;
    private final String baseUrl;
    private final Duration timeout;
    private final HttpClient http = HttpClient.newHttpClient();

    public Aquidify(String apiKey) {
        this(apiKey, "https://api.aquidify.com", Duration.ofSeconds(60));
    }

    public Aquidify(String apiKey, String baseUrl, Duration timeout) {
        String key = apiKey != null ? apiKey : System.getenv("AQUIDIFY_API_KEY");
        if (key == null || key.isEmpty()) {
            throw new IllegalArgumentException("Aquidify API key missing: pass it or set AQUIDIFY_API_KEY.");
        }
        this.apiKey = key;
        this.baseUrl = baseUrl.replaceAll("/+$", "");
        this.timeout = timeout;
    }

    /** Result of an interpretation. {@code clarification} is null unless the input cannot be interpreted without asking. */
    public record Response(String requestId, String domain, String parserVersion, String schemaVersion,
                           JsonNode interpretation, JsonNode clarification, JsonNode meta) {}

    public Response interpret(String domain, String input, String locale) {
        return interpret(domain, input, locale, null, null, null);
    }

    /** Interpret free text into structured intent. Versions and idempotencyKey may be null. */
    public Response interpret(String domain, String input, String locale,
                              String parserVersion, String schemaVersion, String idempotencyKey) {
        ObjectNode body = JSON.createObjectNode().put("domain", domain).put("input", input).put("locale", locale);
        if (parserVersion != null) body.put("parser_version", parserVersion);
        if (schemaVersion != null) body.put("schema_version", schemaVersion);

        JsonNode data = send("POST", "/v1/interpret", body, idempotencyKey);
        return new Response(data.path("request_id").asText(), data.path("domain").asText(),
                data.path("parser_version").asText(), data.path("schema_version").asText(),
                data.path("interpretation"), data.path("clarification").isNull() ? null : data.path("clarification"),
                data.path("meta"));
    }

    /**
     * Register a task version ({@code name@version}). The definition is the task's JSON
     * (description, instructions, schema, examples) as a Map, record or JsonNode. The same
     * definition again is a no-op; a changed one fails with code {@code task_exists}:
     * register a new version instead. Returns the registered task.
     */
    public JsonNode putTask(String id, Object definition) {
        return send("PUT", "/v1/tasks/" + pathSegment(id), JSON.valueToTree(definition), null);
    }

    /** A registered task; code {@code not_found} when this key has none by that id. */
    public JsonNode getTask(String id) {
        return send("GET", "/v1/tasks/" + pathSegment(id), null, null);
    }

    /** The task ids registered with this API key. */
    public List<String> listTasks() {
        List<String> ids = new ArrayList<>();
        send("GET", "/v1/tasks", null, null).path("tasks").forEach(t -> ids.add(t.asText()));
        return ids;
    }

    private static String pathSegment(String id) {
        return URLEncoder.encode(id, StandardCharsets.UTF_8).replace("+", "%20").replace("%40", "@");
    }

    private JsonNode send(String method, String path, JsonNode body, String idempotencyKey) {
        HttpRequest.Builder req = HttpRequest.newBuilder(URI.create(baseUrl + path))
                .timeout(timeout)
                .header("Authorization", "Bearer " + apiKey)
                .header("Accept", "application/json")
                .header("User-Agent", "aquidify-sdk-java/" + VERSION)
                .method(method, body == null ? HttpRequest.BodyPublishers.noBody()
                        : HttpRequest.BodyPublishers.ofString(body.toString()));
        if (body != null) req.header("Content-Type", "application/json");
        if (idempotencyKey != null) req.header("Idempotency-Key", idempotencyKey);

        HttpResponse<String> res;
        try {
            res = http.send(req.build(), HttpResponse.BodyHandlers.ofString());
        } catch (IOException e) {
            throw new AquidifyException("Aquidify request failed: " + e.getMessage(), 0, "network", null, null);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new AquidifyException("Aquidify request interrupted", 0, "network", null, null);
        }

        JsonNode data;
        try {
            data = JSON.readTree(res.body());
        } catch (IOException e) {
            data = JSON.createObjectNode();
        }
        if (data == null || data.isMissingNode()) data = JSON.createObjectNode();
        if (res.statusCode() >= 200 && res.statusCode() < 300) {
            return data;
        }
        Integer retryAfter = res.headers().firstValue("Retry-After").map(v -> {
            try { return Integer.valueOf(v); } catch (NumberFormatException e) { return null; }
        }).orElse(null);
        String requestId = data.hasNonNull("request_id") ? data.get("request_id").asText()
                : res.headers().firstValue("X-Request-ID").orElse(null);
        throw new AquidifyException(
                data.at("/error/message").asText("Aquidify returned HTTP " + res.statusCode()),
                res.statusCode(), data.at("/error/code").asText("unknown"), requestId, retryAfter);
    }

    /** A failed call. {@code code} is the API's stable error code, or "network". */
    public static final class AquidifyException extends RuntimeException {
        private static final Set<String> RETRYABLE = Set.of("rate_limited", "model_unavailable", "timeout", "network");
        public final int status;
        public final String code;
        public final String requestId;
        public final Integer retryAfter;

        AquidifyException(String message, int status, String code, String requestId, Integer retryAfter) {
            super(message);
            this.status = status;
            this.code = code;
            this.requestId = requestId;
            this.retryAfter = retryAfter;
        }

        public boolean isRetryable() { return RETRYABLE.contains(code); }
    }
}

File: go/aquidify.go
// Package aquidify is the Go client for the Aquidify API: turn what people type
// into validated, structured intent.
//
//	c, _ := aquidify.New("")  // reads AQUIDIFY_API_KEY
//	r, err := c.Interpret(ctx, aquidify.Request{
//		Domain: "hiring.candidate",
//		Input:  "Iščem delo v skladišču v Ljubljani, brez nočnih.",
//		Locale: "sl-SI",
//	})
//
// Interpretation is returned as raw JSON; unmarshal it into your own types or
// see schemas/hiring.candidate-1.0.0.json in the repository.
package aquidify

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"os"
	"strconv"
	"strings"
	"time"
)

const Version = "0.1.0"

type Request struct {
	Domain        string `json:"domain"`
	Input         string `json:"input"`
	Locale        string `json:"locale"`
	ParserVersion string `json:"parser_version,omitempty"`
	SchemaVersion string `json:"schema_version,omitempty"`
}

type Response struct {
	RequestID      string          `json:"request_id"`
	Domain         string          `json:"domain"`
	ParserVersion  string          `json:"parser_version"`
	SchemaVersion  string          `json:"schema_version"`
	Interpretation json.RawMessage `json:"interpretation"`
	Clarification  *Clarification  `json:"clarification"`
	Meta           Meta            `json:"meta"`
}

type Clarification struct {
	Reason           string   `json:"reason"`
	Field            string   `json:"field"`
	QuestionKey      string   `json:"question_key"`
	SuggestedOptions []string `json:"suggested_options"`
}

type Meta struct {
	Source    string `json:"source"` // deterministic | ai
	Cache     string `json:"cache"`  // bypass | hit | miss | coalesced
	Provider  string `json:"provider,omitempty"`
	Model     string `json:"model,omitempty"`
	LatencyMS int64  `json:"latency_ms"`
}

// Error is a failed call. Code is the API's stable error code, or "network".
type Error struct {
	Status     int
	Code       string
	Message    string
	RequestID  string
	RetryAfter time.Duration
}

func (e *Error) Error() string {
	return fmt.Sprintf("aquidify: %d %s: %s", e.Status, e.Code, e.Message)
}

// Retryable reports whether the same request may succeed later.
func (e *Error) Retryable() bool {
	switch e.Code {
	case "rate_limited", "model_unavailable", "timeout", "network":
		return true
	}
	return false
}

type Client struct {
	apiKey  string
	BaseURL string
	HTTP    *http.Client
}

// New creates a client. An empty apiKey reads AQUIDIFY_API_KEY.
func New(apiKey string) (*Client, error) {
	if apiKey == "" {
		apiKey = os.Getenv("AQUIDIFY_API_KEY")
	}
	if apiKey == "" {
		return nil, errors.New("aquidify: API key missing: pass it or set AQUIDIFY_API_KEY")
	}
	return &Client{apiKey: apiKey, BaseURL: "https://api.aquidify.com", HTTP: &http.Client{Timeout: 60 * time.Second}}, nil
}

// Option adjusts a single call.
type Option func(*http.Request)

// WithIdempotencyKey replays the first response for the same key and body for 24h.
func WithIdempotencyKey(key string) Option {
	return func(r *http.Request) { r.Header.Set("Idempotency-Key", key) }
}

// Interpret turns free text into structured intent.
func (c *Client) Interpret(ctx context.Context, req Request, opts ...Option) (*Response, error) {
	body, err := json.Marshal(req)
	if err != nil {
		return nil, err
	}
	hr, err := http.NewRequestWithContext(ctx, http.MethodPost, strings.TrimRight(c.BaseURL, "/")+"/v1/interpret", bytes.NewReader(body))
	if err != nil {
		return nil, err
	}
	hr.Header.Set("Authorization", "Bearer "+c.apiKey)
	hr.Header.Set("Content-Type", "application/json")
	hr.Header.Set("Accept", "application/json")
	hr.Header.Set("User-Agent", "aquidify-sdk-go/"+Version)
	for _, o := range opts {
		o(hr)
	}

	res, err := c.HTTP.Do(hr)
	if err != nil {
		return nil, &Error{Code: "network", Message: err.Error()}
	}
	defer res.Body.Close()
	raw, err := io.ReadAll(io.LimitReader(res.Body, 4<<20))
	if err != nil {
		return nil, &Error{Status: res.StatusCode, Code: "network", Message: err.Error()}
	}

	if res.StatusCode >= 200 && res.StatusCode < 300 {
		var out Response
		if err := json.Unmarshal(raw, &out); err != nil {
			return nil, fmt.Errorf("aquidify: decode response: %w", err)
		}
		return &out, nil
	}

	var e struct {
		Error struct {
			Code    string `json:"code"`
			Message string `json:"message"`
		} `json:"error"`
		RequestID string `json:"request_id"`
	}
	_ = json.Unmarshal(raw, &e)
	apiErr := &Error{Status: res.StatusCode, Code: e.Error.Code, Message: e.Error.Message, RequestID: e.RequestID}
	if apiErr.Code == "" {
		apiErr.Code, apiErr.Message = "unknown", fmt.Sprintf("HTTP %d", res.StatusCode)
	}
	if apiErr.RequestID == "" {
		apiErr.RequestID = res.Header.Get("X-Request-ID")
	}
	if s, err := strconv.Atoi(res.Header.Get("Retry-After")); err == nil {
		apiErr.RetryAfter = time.Duration(s) * time.Second
	}
	return nil, apiErr
}
